### PR TITLE
[#1103] Extend the Statemachine Xtext Editor folding capabilities.

### PR DESCRIPTION
**Pull request by miklossy**
- Provide the StatemachineFoldingRegionProvider custom class that
computes folding regions not only for the states, but also for the
events, resetEvent and commands blocks.
- Provide corresponding binding in the StatemachineUiModule.
- Provide corresponding StatemachineFoldingTest UI test cases.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	private def ITextRegion textRegionBetween(XtextResource it, Keyword startKeyword, Keyword endKeyword) {
+		var INode startNode
+		for (INode node : parseResult.rootNode.asTreeIterable) {
+			val grammarElement = node.grammarElement
+			if (grammarElement == startKeyword) {
+				startNode = node
+			}
+			if (grammarElement == endKeyword && startNode !==null) {
+				val endNode = node
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend]: Why is this local variable necessary?

File: org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			}
+		}
+	}
+
+	/**
+	 * Determine the text region between the start node and the end node.
+	 */
+	private def ITextRegion textRegionBetween(XtextResource it, INode startNode, INode endNode) {
+		val offset = startNode.offset
+		val length = endNode.offset + endNode.length - startNode.offset -1
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend]: You may want to use `endNode.endOffset`

Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend]: What is the -1 about? To me it indicates a problem somewhere else in the logic.

Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.ui/src/org/eclipse/xtext/example/fowlerdsl/ui/folding/StatemachineFoldingRegionProvider.xtend]: `startNode.offset` can be replaced by local var `offset` declared a line above

**Comment by miklossy** [comment]
Thanks for the review @szarnekow ! I updated the PR as requested.

**Pull request merged**